optaweb-employee-rostering-frontend/src/store/contract/selectors.ts

```diff
@@ -24,20 +25,19 @@ export const getContractById = (state: AppState, id: number): Contract => {
 };
 
 let oldContractMapById: Map<number, Contract> | null = null;
-let contractListForOldContractMapById: Contract[] | null = null;
+let contractListForOldContractMapById: List<Contract> | null = null;
 
-export const getContractList = (state: AppState): Contract[] => {
+export const getContractList = (state: AppState): List<Contract> => {
   if (state.contractList.isLoading) {
-    return [];
+    return List();
   }
   if (oldContractMapById === state.contractList.contractMapById && contractListForOldContractMapById !== null) {
     return contractListForOldContractMapById;
   }
-
-  const out: Contract[] = [];
-  state.contractList.contractMapById.forEach((value, key) => out.push(getContractById(state, key)));
+  const out = state.contractList.contractMapById.keySeq().map(id => getContractById(state, id)).toList();
 
   oldContractMapById = state.contractList.contractMapById;
```

Comment:
IDEA says:
>Assigned expression type Map<number, DomainObjectView<Contract>> is not assignable to type Map<number, Contract> | null 
which kinda makes sense since
```ts
let oldContractMapById: Map<number, Contract> | null = null;
```
versus
```ts
readonly contractMapById: Map<number, DomainObjectView<Contract>>;
```
I wonder why `npm run typecheck` doesn't fail on this.

Or am I missing something?

Reply:
`DomainObjectView<Contract> = Contract` since `Contract` does not refer to any other domain objects (such as skill, spot, etc.). `DomainObjectView<Employee> != Employee`, since Employee refers to `Skill` and `Contract` (which will be replaced by numbers in their definitions. 

Now that https://github.com/microsoft/TypeScript/issues/12754 is fixed via https://github.com/microsoft/TypeScript/pull/40336, if and when we update to a new typescript release, I can remove the two different definitions of DomainObjectViews in the store (which has to deal with converting spotId -> spot), which would make this a type error.